File: TheepotAndroid/app/src/main/kotlin/nl/bsodetheepot/mobile/data/models/Taken.kt
package nl.bsodetheepot.mobile.data.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import nl.bsodetheepot.mobile.data.services.DateUtils

@Serializable
enum class LijstType {
    @SerialName("taken") TAKEN,
    @SerialName("notities") NOTITIES,
}

enum class Prioriteit(val waarde: Int) {
    GEEN(0), LAAG(1), MEDIUM(2), HOOG(3);

    val label: String
        get() = when (this) {
            GEEN -> "Geen"
            LAAG -> "Laag"
            MEDIUM -> "Medium"
            HOOG -> "Hoog"
        }

    val kleurHex: String
        get() = when (this) {
            GEEN -> "#9CA3AF"
            LAAG -> "#3B82F6"
            MEDIUM -> "#F59E0B"
            HOOG -> "#EF4444"
        }

    companion object {
        fun vanWaarde(w: Int): Prioriteit = entries.firstOrNull { it.waarde == w } ?: GEEN
    }
}

/** Tabel `todo_lijsten`. */
@Serializable
data class TodoLijst(
    val id: String,
    val naam: String,
    val kleur: String,
    @SerialName("eigenaar_id") val eigenaarId: String,
    val volgorde: Int,
    val type: LijstType,
)

/** Tabel `todo_taken`. */
@Serializable
data class TodoTaak(
    val id: String,
    @SerialName("lijst_id") val lijstId: String,
    val titel: String,
    val notitie: String? = null,
    val voltooid: Boolean = false,
    val prioriteit: Int = 0,
    val vervaldatum: String? = null,
    val volgorde: Int = 0,
    @SerialName("voltooid_op") val voltooidOp: String? = null,
) {
    val prioriteitEnum: Prioriteit get() = Prioriteit.vanWaarde(prioriteit)

    val isVerlopen: Boolean
        get() {
            if (voltooid) return false
            val datum = vervaldatum?.let { DateUtils.parseDateStr(it) } ?: return false
            return datum.isBefore(DateUtils.vandaag())
        }

    val isVandaag: Boolean
        get() = vervaldatum == DateUtils.toDateStr(DateUtils.vandaag())
}

/** Tabel `notities`. */
@Serializable
data class Notitie(
    val id: String,
    @SerialName("lijst_id") val lijstId: String,
    val titel: String,
    val inhoud: String,
    val kleur: String = "#ffffff",
    val volgorde: Int = 0,
    @SerialName("bijgewerkt_op") val bijgewerktOp: String? = null,
)

File: TheepotMobile/TheepotMobile/Models/Taken.swift
import Foundation

enum TodoLijstType: String, Codable {
    case taken, notities
}

/// Tabel `todo_lijsten`.
struct TodoLijst: Codable, Identifiable, Hashable {
    let id: String
    var naam: String
    var kleur: String
    let eigenaarId: String
    var volgorde: Int
    let type: TodoLijstType

    enum CodingKeys: String, CodingKey {
        case id, naam, kleur
        case eigenaarId = "eigenaar_id"
        case volgorde, type
    }
}

enum Prioriteit: Int, Codable, CaseIterable {
    case geen = 0, laag = 1, medium = 2, hoog = 3

    var label: String {
        switch self {
        case .geen: return "Geen"
        case .laag: return "Laag"
        case .medium: return "Medium"
        case .hoog: return "Hoog"
        }
    }

    var kleurHex: String {
        switch self {
        case .geen: return "#9CA3AF"
        case .laag: return "#3B82F6"
        case .medium: return "#F59E0B"
        case .hoog: return "#EF4444"
        }
    }
}

/// Tabel `todo_taken`.
struct TodoTaak: Codable, Identifiable, Hashable {
    let id: String
    let lijstId: String
    var titel: String
    var notitie: String?
    var voltooid: Bool
    var prioriteit: Prioriteit
    var vervaldatum: String?
    var volgorde: Int
    var voltooidOp: String?

    enum CodingKeys: String, CodingKey {
        case id
        case lijstId = "lijst_id"
        case titel, notitie, voltooid, prioriteit, vervaldatum, volgorde
        case voltooidOp = "voltooid_op"
    }

    var isVerlopen: Bool {
        guard !voltooid, let vervaldatum else { return false }
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        guard let datum = f.date(from: vervaldatum) else { return false }
        return datum < Calendar.current.startOfDay(for: Date())
    }

    var isVandaag: Bool {
        guard let vervaldatum else { return false }
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return vervaldatum == f.string(from: Date())
    }
}

/// Tabel `notities`.
struct Notitie: Codable, Identifiable {
    let id: String
    let lijstId: String
    var titel: String
    var inhoud: String
    var kleur: String
    var volgorde: Int
    var bijgewerktOp: String

    enum CodingKeys: String, CodingKey {
        case id
        case lijstId = "lijst_id"
        case titel, inhoud, kleur, volgorde
        case bijgewerktOp = "bijgewerkt_op"
    }
}
